File: sol/solution/0101-0200/0109/Solution.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-spec sorted_list_to_bst(Head :: #list_node{} | null) -> #tree_node{} | null.
sorted_list_to_bst(null) ->
    null;
sorted_list_to_bst(Head) ->
    List = to_list(Head),
    build_bst(List).
to_list(null) -> [];
to_list(#list_node{val = V, next = Next}) ->
    [V | to_list(Next)].
build_bst([]) -> null;
build_bst(List) ->
    Len = length(List),
    Mid = Len div 2,
    {Left, [Root|Right]} = lists:split(Mid, List),
    #tree_node{
        val = Root,
        left = build_bst(Left),
        right = build_bst(Right)
    }.

File: sol/solution/0101-0200/0109/Solution.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end
defmodule Solution do
  @spec sorted_list_to_bst(head :: ListNode.t() | nil) :: TreeNode.t() | nil
  def sorted_list_to_bst(head) do
    list = to_array(head)
    build_bst(list, 0, length(list) - 1)
  end
  defp to_array(nil), do: []
  defp to_array(%ListNode{val: val, next: next}) do
    [val | to_array(next)]
  end
  defp build_bst(list, left, right) when left > right, do: nil
  defp build_bst(list, left, right) do
    mid = div(left + right, 2)
    val = Enum.at(list, mid)
    %TreeNode{
      val: val,
      left: build_bst(list, left, mid - 1),
      right: build_bst(list, mid + 1, right)
    }
  end
end
